servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp

```diff
@@ -1364,11 +1376,35 @@ void RenderForwardClustered::_pre_opaque_render(RenderDataRD *p_render_data, boo
 		}
 	}
 
-	// Render GI
+	// Render GI.
 
-	bool render_shadows = p_render_data->directional_shadows.size() || p_render_data->shadows.size();
+	bool render_static_shadows = static_cube_shadows.size() > 0 || static_positional_shadows.size() > 0;
+	bool render_shadows = render_static_shadows || p_render_data->directional_shadows.size() || p_render_data->shadows.size() || p_render_data->cube_shadows.size();
 	bool render_gi = rb.is_valid() && p_use_gi;
 
+	// Prepare shadow rendering.
+	if (render_static_shadows) {
+		RENDER_TIMESTAMP("Render Static Shadows");
+
+		// Cube shadows are rendered in their own way so we do them first.
+		for (uint32_t i = 0; i < static_cube_shadows.size(); i++) {
```

Comment:
The fact that this is the only explicit cycle in commit concers me (except the same cycle in render_forward_mobile.h). Could that be the CPU bottleneck that @Calinou has found through minimal reproduction project?
@BastiaanOlij @DarioSamo please investigate. Is it possible to parallelize such cycle? I am bypasser with ~0% programming experience and the fact we have nested loop through frame rendering makes me wonder if that's necessary.

Reply:
Parallelizing is a very dangerous option to consider when it concerns draw lists, as there's no possible way of appending contents to a single command buffer from multiple threads without relying on secondary command lists. And secondary command lists have very dodgy support across the board (NVIDIA recently fixed a bug related to them but I imagine YMMV on Android).